java_lang_invoke_MethodHandlesProxiesTest failure on jdk8u

Failure link

https://ci.adoptopenjdk.net/job/openjdk8_j9_openjdktest_x86-64_linux/547/testReport/

    test category: openjdk
    test suite name: jdk_lang
    OS/architecture: fails on all platforms

java -version
05:42:04  openjdk version "1.8.0_222"
05:42:04  OpenJDK Runtime Environment (build 1.8.0_222-201905290400-b04)
05:42:04  Eclipse OpenJ9 VM (build master-501661ce5, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20190529_322 (JIT enabled, AOT enabled)
05:42:04  OpenJ9   - 501661ce5
05:42:04  OMR      - b2ffe8a9
05:42:04  JCL      - fc2cdb9145 based on jdk8u222-b04)
11:23:42 JCL - bb38f8aaac based on jdk-12+32)

Optional info

    intermittent failure: no
    regression: yes, started failing around May 22nd, according to test results 'Age'

Failure output

05:49:52  [TestNG] Running:
05:49:52    java/lang/invoke/MethodHandlesProxiesTest.java
05:49:52  
05:49:52  test test.java.lang.invoke.MethodHandlesProxiesTest.testDefaultMethods(): failure
05:49:52  java.lang.InternalError: java.lang.IllegalAccessException: test/java/lang/invoke/MethodHandlesProxiesTest$A is not a direct superinterface of com/sun/proxy/$Proxy2.
05:49:52  	at java.lang.invoke.MethodHandleProxies.lambda$callDefaultMethod$0(MethodHandleProxies.java:367)
05:49:52  	at java.lang.invoke.MethodHandleProxies$$Lambda$2.000000001C17D990.apply(Unknown Source)
05:49:52  	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
05:49:52  	at java.lang.invoke.MethodHandleProxies.callDefaultMethod(MethodHandleProxies.java:354)
05:49:52  	at java.lang.invoke.MethodHandleProxies.access$300(MethodHandleProxies.java:44)
05:49:52  	at java.lang.invoke.MethodHandleProxies$1.invoke(MethodHandleProxies.java:200)
05:49:52  	at com.sun.proxy.$Proxy2.a(Unknown Source)
05:49:52  	at test.java.lang.invoke.MethodHandlesProxiesTest.testDefaultMethods(MethodHandlesProxiesTest.java:87)
05:49:52  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
05:49:52  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
05:49:52  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
05:49:52  	at java.lang.reflect.Method.invoke(Method.java:498)
05:49:52  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
05:49:52  	at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
05:49:52  	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:821)
05:49:52  	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1131)
05:49:52  	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
05:49:52  	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
05:49:52  	at org.testng.TestRunner.privateRun(TestRunner.java:773)
05:49:52  	at org.testng.TestRunner.run(TestRunner.java:623)
05:49:52  	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
05:49:52  	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
05:49:52  	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
05:49:52  	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
05:49:52  	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
05:49:52  	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
05:49:52  	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1185)
05:49:52  	at org.testng.TestNG.runSuitesLocally(TestNG.java:1110)
05:49:52  	at org.testng.TestNG.run(TestNG.java:1018)
05:49:52  	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:94)
05:49:52  	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:54)
05:49:52  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
05:49:52  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
05:49:52  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
05:49:52  	at java.lang.reflect.Method.invoke(Method.java:498)
05:49:52  	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
05:49:52  	at java.lang.Thread.run(Thread.java:819)
05:49:52  Caused by: java.lang.IllegalAccessException: test/java/lang/invoke/MethodHandlesProxiesTest$A is not a direct superinterface of com/sun/proxy/$Proxy2.
05:49:52  	at java.lang.invoke.PrimitiveHandle.finishMethodInitialization(PrimitiveHandle.java:196)
05:49:52  	at java.lang.invoke.DirectHandle.<init>(DirectHandle.java:51)
05:49:52  	at java.lang.invoke.DirectHandle.<init>(DirectHandle.java:43)
05:49:52  	at java.lang.invoke.MethodHandles$Lookup.findSpecialImpl(MethodHandles.java:512)
05:49:52  	at java.lang.invoke.MethodHandles$Lookup.findSpecial(MethodHandles.java:480)
05:49:52  	at java.lang.invoke.MethodHandleProxies.lambda$callDefaultMethod$0(MethodHandleProxies.java:358)
05:49:52  	... 36 more
05:49:52  test test.java.lang.invoke.MethodHandlesProxiesTest.testOverriddenDefaultMethods(): success
05:49:52  
05:49:52  ===============================================
05:49:52  java/lang/invoke/MethodHandlesProxiesTest.java
05:49:52  Total tests run: 2, Failures: 1, Skips: 0
05:49:52  ===============================================
05:49:52  
05:49:52  STDERR:
05:49:52  java.lang.Exception: failures: 1
05:49:52  	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:96)
05:49:52  	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:54)
05:49:52  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
05:49:52  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
05:49:52  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
05:49:52  	at java.lang.reflect.Method.invoke(Method.java:498)
05:49:52  	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
05:49:52  	at java.lang.Thread.run(Thread.java:819)
05:49:52  
05:49:52  JavaTest Message: Test threw exception: java.lang.Exception: failures: 1
05:49:52  JavaTest Message: shutting down test

--- a/runtime/jcl/common/proxy.c
+++ b/runtime/jcl/common/proxy.c
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1998, 2018 IBM Corp. and others
+ * Copyright (c) 1998, 2019 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -23,6 +23,23 @@
 #include "jclglob.h"
 #include "jclprots.h"
 
+#if JAVA_SPEC_VERSION == 8
+static BOOLEAN isProxyExemptFromVerification(JNIEnv * env, jstring className) {
+	BOOLEAN result = FALSE;
+	if (NULL != className) {
+		const char * utf8Name = (const char *) (*env)->GetStringUTFChars(env, className, NULL);
+		if (NULL != utf8Name) {
+			/* check if class name matches the generated implementation class name excluding the generated characters */
+			if ((0 == strncmp(utf8Name, "com.sun.proxy.$Proxy", 20))) {
+				result = TRUE;
+			}
+			(*env)->ReleaseStringUTFChars(env, className, utf8Name);
+		}
+	}
+	return result;
+}
+#endif /* JAVA_SPEC_VERSION == 8 */
+
 static jclass proxyDefineClass(
 		JNIEnv * env,
 		jobject classLoader,
@@ -37,6 +54,7 @@ static jclass proxyDefineClass(
 	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
 	J9StackWalkState walkState;
 	J9Class * clazz = NULL;
+	UDATA defineClassOptions = 0;
 
 	/* Walk the stack to find the caller class loader and protection domain */
 	vmFuncs->internalEnterVMFromJNI(currentThread);
@@ -65,7 +83,12 @@ static jclass proxyDefineClass(
 
 	vmFuncs->internalExitVMToJNI(currentThread);
 
-	return defineClassCommon(env, classLoader, className, classBytes, offset, length, pd, 0, NULL);
+#if JAVA_SPEC_VERSION == 8
+	if (isProxyExemptFromVerification(env, className)) {
+		defineClassOptions |= J9_FINDCLASS_FLAG_UNSAFE;
+	}
+#endif /* JAVA_SPEC_VERSION == 8 */
+	return defineClassCommon(env, classLoader, className, classBytes, offset, length, pd, defineClassOptions, NULL);
 }
 
 jclass JNICALL Java_java_lang_reflect_Proxy_defineClassImpl(JNIEnv * env, jclass recvClass, jobject classLoader, jstring className, jbyteArray classBytes)
@@ -90,7 +113,13 @@ Java_java_lang_reflect_Proxy_defineClass0__Ljava_lang_ClassLoader_2Ljava_lang_St
 	if (classLoader == NULL || pd == NULL) {
 		return proxyDefineClass(env, classLoader, className, classBytes, offset, length, pd);
 	} else {
-		return defineClassCommon(env, classLoader, className, classBytes, offset, length, pd, 0, NULL);
+		UDATA defineClassOptions = 0;
+#if JAVA_SPEC_VERSION == 8
+		if (isProxyExemptFromVerification(env, className)) {
+			defineClassOptions |= J9_FINDCLASS_FLAG_UNSAFE;
+		}
+#endif /* JAVA_SPEC_VERSION == 8 */
+		return defineClassCommon(env, classLoader, className, classBytes, offset, length, pd, defineClassOptions, NULL);
 	}
 }
 

